Hack'n'Hustle: Support rebasing images from specified Git commit hashes

vfreex opened the pull request:
This PR adds a new option `--upstream` to the `images:rebase` command so that Doozer can build images from specified Git commits.
This would give us the ability to rebuilding an image any time without introducing unwanted code changes.

To specify the commit hashes, use this syntax:

```bash
doozer images:rebase --upstream distgit_key_1=commitish_1 --upstream
distgit_key_2=commitish_2 ...
```

For example:

```bash
doozer -g openshift-4.4 --images ose-metering-ansible-operator:c0d9aaf5a901cc9b40068cac0b9c037579d72dc7,openshift-enterprise-ansible-operator \
    --latest-parent-version images:rebase -m 'test' --version v4.4 --release 202001010000 --upstream ose-metering-ansible-operator=c0d9aaf5a901cc9b40068cac0b9c037579d72dc7
```

Note you images specified in `--images` without `--upstream` will use
the tip commit of the branch so it will not break the backward
compatibility.

Limitations:
- Building RPMs from commit hashes is not covered in this PR.
- The behavior of `--latest-parent-version` is not changed, so it could use a base image from the latest Brew build.
- commit-ish format is not checked, so it relies on `git` to check that.

In the future we can also have an option like `--parent` with the same
syntax to allow us to
specify parent image NVRs.

sosiouxme commented on doozerlib/distgit.py:
```diff
@@ -1601,6 +1601,10 @@ def _merge_source(self):
 
         with Dir(self.source_path()):
+            if self.metadata.commitish:
+                self.runtime.logger.info(f"Rebasing image {self.name} from speicified commit-ish {self.metadata.commitish}...")
```
```suggestion
                self.runtime.logger.info(f"Rebasing image {self.name} from specified commit-ish {self.metadata.commitish}...")
```

sosiouxme commented on doozerlib/runtime.py:
```diff
@@ -398,7 +398,19 @@ def filter_disabled(n, d):
+                if len(image_split) < 2:
+                    continue
+                commitish = image_split[1]
+                # TODO: verify `commitish` format
```
meh... I think it's fine not to.

vfreex replied:
Well, now I think we'd better not to implement that check on Doozer. Doozer should be able to build from anywhere. We can implement the check in the component invoking Doozer, like a Jenkins job.

sosiouxme commented:
BTW we need this for RPMs too. Not as important, but I think we'll have a use case.

vfreex commented:
@sosiouxme RPM will be covered... Though I have not come up with a plan to rebuild images after that RPM rebuilds... Maybe we will link RPMs to images in the fuy

jupierce commented:
Mainly a UI observation: pairing the commit with the -i argument makes me think you are specifying a distgit commit. The -i entries are effectively distgit identifiers.
`--upstream github.com....:hash` would be clearer to me. If a distgit has an upstream that matches a specified upstream, it would use that hash instead of the default.

vfreex commented:
@jupierce Makes sense. I will update later.

vfreex commented:
@jupierce I updated the code with following changes:

- I moved the hash specification to dedicated `--upstream` option using syntax `--upstream some-image=hash`. But it still uses distgit names as the identifiers. One consideration is we share sources for some images using "alias". If we override the hash when a distgit's upstream matches  "github.com...", it will override all images sharing that source. Another thought is that in the near future we may provide the input via Jenkins build parameter. It could be easier for people to use the distgit names everywhere, otherwise people have to lookup ocp-build-data for the source URLs in order to fill the parameter. Considering the option is named to `upstream`, I think it becomes more clear than reuse the `--images` option, and gives us the flexibility to only override the commits for a subset of images.
- https://github.com/openshift/doozer/pull/201 introduced a behavior change that shallow clone is used for better clone performance. I added an additional step to fetch the commit explicitly after git clone.

PTAL. Thanks.

vfreex commented:
@jupierce @sosiouxme Can we merge this or you guys still have suggestions?
